### ISSUE TITLE
Are UPSERT statements DML statements?

### ISSUE THREAD

**Issue by hoehrmann** [coleifer/pysqlite3#4]
Per https://docs.python.org/3/library/sqlite3.html#controlling-transactions »The Python sqlite3 module by default issues a BEGIN statement implicitly before a Data Modification Language (DML) statement (i.e. INSERT/UPDATE/DELETE/REPLACE).«

Associated code seems to be https://github.com/coleifer/pysqlite3/blob/c38df85c90115fb7cb00565c4154aeaf92b67184/src/statement.c#L76

I do not quite understand how statements preceded by comments or statements that had not been known to be supported by SQLite when that code was written are handled, like https://sqlite.org/lang_UPSERT.html which is a mix of INSERT and UPDATE but seemingly is not considered by the code linked above.

When my code uses `UPSERT`, but the current version of the library does not consider `UPSERT` to be a modifying statement, I could work around that by starting a transaction manually. But a later version of this library might add support for `UPSERT`, in which case my workaround breaks, at runtime, depending on which version is installed.

**Comment by coleifer** [comment]
Upsert is not a statement. If you had actually used upsert you would realize this.

**Comment by hoehrmann** [comment]
@coleifer: That is entirely irrelevant here, but I take it you are not qualified to address the underlying problem anyways, the Python documentation rather refers to https://github.com/ghaering/pysqlite which has not been updated in years, so I will raise this issue with qualified people instead.

**Comment by coleifer** [comment]
Sqlite does not offer an "UPSERT" statement. Upsert occurs via REPLACE *or* using the new postgres-compatible INSERT ... ON CONFLICT DO UPDATE, in which case it will match INSERT.

**Comment by hoehrmann** [comment]
Like I said, that is irrelevant to the issue I raised.

**Comment by coleifer** [comment]
If your statement is preceded by comments or uses a CTE (I actually just [opened a bug on python's tracker](https://bugs.python.org/issue36859), prompted by your issue, so hopefully something comes of that upstream, if I don't patch it here), then pysqlite3 (which is based on stdlib sqlite3) will incorrectly assume it is *not* a DML statement.

The solution, in my own experience, is to completely ditch the *absolutely insane* transaction handling semantics of pysqlite and manage my own transactions.

To do this, you set the isolation_level of the conn to `None`.

```python
from pysqlite3 import dbapi2 as sqlite3
# or, import sqlite3

conn = sqlite3.connect('foo.db', isolation_level=None)

# Or,
conn = sqlite3.connect('foo.db')
conn.isolation_level = None
```

Now your conn is in autocommit mode. You can control the transactions 100% by issuing begin statements and then calling conn.commit() / conn.rollback().

More info described in a blog post I wrote on the matter: http://charlesleifer.com/blog/going-fast-with-sqlite-and-python/

**Comment by coleifer** [comment]
Demonstration of the bug/behavior I posted on the python bug tracker:

```python
import sqlite3

conn = sqlite3.connect(':memory:')

conn.execute('create table kv ("key" text primary key, "value" integer)')
conn.execute('insert into kv (key, value) values (?, ?), (?, ?)',
             ('k1', 1, 'k2', 2))

assert conn.in_transaction  # Yes we are in a transaction.
conn.commit()
assert not conn.in_transaction  # Not anymore, as expected.

rc = conn.execute('update kv set value=value + ?', (10,))
print(rc.rowcount)
assert conn.in_transaction
conn.commit()
assert not conn.in_transaction

rc = conn.execute(
    'with c(k, v) as (select key, value + 10 from kv) '
    'update kv set value=(select v from c where k=kv.key)')

print(rc.rowcount)  # Should be 2, prints "-1".
#assert conn.in_transaction  # !!! Fails.

curs = conn.execute('select * from kv order by key;')
print(curs.fetchall())
```

**Comment by coleifer** [comment]
PS: UPSERT is not a statement.

**Comment by hoehrmann** [comment]
@coleifer: This is looking good, thank you. I was coming from "the documentation says 'INSERT/UPDATE/DELETE/REPLACE', but what if SQLite adds another command XYZ" and mistakenly googled »UPSERT« as a good example for a recently added command. You correctly pointed out that this was not a good example. Looking at the code, I also noticed that comments are not skipped either, so this was clearly a problem, but I filed the issue prematurely and in the wrong place, sorry for that, and thanks for your initiative filing the bug!

**Comment by coleifer** [comment]
If this weren't intended to be a standalone package of the stdlib, I'd drop the bogus transaction stuff and go closer to sqlite's own apis (sqlite3_get_autocommit). Setting isolation_level to `None` (which must be done explicitly) is an easy enough way to get around it.

For a different package, check out rogerbinns/apsw, close to c, full API implementation, no bullshit.

**Comment by coleifer** [comment]
I've patched this repo with 65a9d4e1dbcbcca637b534cae25394ff2a09b4b7 -- and have submitted a more robust patch (adds support for sqlite prior to 3.7.11) to the python issue tracker. We'll see what happens I guess, but it should be much more reliable at least in this repo.

https://bugs.python.org/issue36859

**Comment by coleifer** [comment]
Official PR is at: python/cpython#13216
